userspace/libsinsp/test/plugins/syscall_source.cpp

```diff
@@ -0,0 +1,166 @@
+/*
+Copyright (C) 2023 The Falco Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+#include <cstring>
+#include <cstdio>
+#include <cstdlib>
+#include <sstream>
+
+#include <ppm_events_public.h>
+
+#include "test_plugins.h"
+
+typedef struct plugin_state
+{
+    std::string lasterr;
+} plugin_state;
+
+typedef struct instance_state
+{
+    uint64_t count;
+    uint8_t evt_buf[2048];
+    ss_plugin_event* evt;
+} instance_state;
+
+static const char* plugin_get_required_api_version()
+{
+    return PLUGIN_API_VERSION_STR;
+}
+
+static const char* plugin_get_version()
+{
+    return "0.1.0";
+}
+
+static const char* plugin_get_name()
+{
+    return "sample_syscall_source";
+}
+
+static const char* plugin_get_description()
+{
+    return "some desc";
+}
+
+static const char* plugin_get_contact()
+{
+    return "some contact";
+}
+
+static const char* plugin_get_last_error(ss_plugin_t* s)
+{
+    return ((plugin_state *) s)->lasterr.c_str();
+}
+
+static ss_plugin_t* plugin_init(const char* config, ss_plugin_rc* rc)
+{
+    *rc = SS_PLUGIN_SUCCESS;
+    return new plugin_state();
+}
+
+static void plugin_destroy(ss_plugin_t* s)
+{
+    delete ((plugin_state *) s);
+}
+
+static ss_instance_t* plugin_open(ss_plugin_t* s, const char* params, ss_plugin_rc* rc)
+{
+    instance_state *ret = new instance_state();
+    ret->evt = (ss_plugin_event*) &ret->evt_buf;
+    ret->count = 10000;
+    auto count = atoi(params);
+    if (count > 0)
+    {
+        ret->count = (uint64_t) count;
+    }
+
+    *rc = SS_PLUGIN_SUCCESS;
+    return ret;
+}
+
+static void plugin_close(ss_plugin_t* s, ss_instance_t* i)
+{
+    delete ((instance_state *) i);
+}
+
+static ss_plugin_rc plugin_next_batch(ss_plugin_t* s, ss_instance_t* i, uint32_t *nevts, ss_plugin_event ***evts)
+{
+    instance_state *istate = (instance_state *) i;
+
+    if (istate->count == 0)
+    {
+        *nevts = 0;
+        return SS_PLUGIN_EOF;
+    }
+
+    *nevts = 1;
+    *evts = &istate->evt;
+    istate->evt->type = PPME_SYSCALL_OPEN_X;
+    istate->evt->tid = 1;
+    istate->evt->ts = UINT64_MAX;
+    istate->evt->len = sizeof(ss_plugin_event);
+    istate->evt->nparams = 6;
+
+    uint8_t* parambuf = &istate->evt_buf[0] + sizeof(ss_plugin_event);
+
+    // lenghts
```

Comment:
Is the SDK going to provide helpers for this? 
Or do we expect that users wanting to send syscall events from a plugin know what they're doing (aka the contract, aka the event table and scap internals)?

Reply:
I see the scap specific to become more open and available to developers in the future now that's part of the plugin system, but of course the SDKs will take care of doing the extra lift in that case. From the Go SDK for example, I expect the UX to not change at all.